cpp/src/tsne/sparse_svd.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include "../../src_prims/utils.h"
+#include "utils.h"
+#include "unary_op.h"
+#include "sparse/coo.h"
+
+
+#define CHECK(x) ASSERT(x == 0, "Failed at line = %d file = %s\n", __LINE__, __FILE__)
+#define MIN(a, b) ((a) > (b)) ? (b) : (a)
+
+
+namespace ML {
+
+
+/**
+ * @brief Randomized SVD on COO sparse matrices
+ * @input param COO_Matrix: Pointer to COO sparse matrix.
+ * @output param U: (n, n_components)
+ * @output param S: (n_components)
+ * @output param VT: (p, n_components)
+ * @input param handle: cuML general GPU handle
+ * @input n_components: How many SVD singular values / singular values you want (ie 2)
+ * @input oversamples: Generally 10 or so. Too few will cause instability.
+ * @input max_iter: Generally 7 or so. Minimum is 3 for stability.
+ * @input random_state: Seed for random normal initialization. -1 is any seed.
+ */
+template <typename math_t> void
+SparseSVD(const MLCommon::Sparse::COO<math_t> *__restrict COO_Matrix,
```

Comment:
Did you test your code with the existing PCA or SVD? Don't know the sparse version but randomized SVD is slower than eig based methods. That's why we don't use the randomized SVD right now even though we have implemented it. Also, randomized SVD is a separate function. It should be included in a separate PR. It doesn't belong to just tSNE.

Reply:
Oh good point. I'll test it on the existing PCA, SVD.
Most likely the sparse version will either be faster (if NNZ(X) is very small) or slower (if NNZ(X) is close to 100%). But ye you're right in saying in general SYEVD / SYEVJ is faster for say n_columns <= 2000 or so.

When n_columns blows up, Randomized SVD wins.

But I can test to confirm :)

Reply:
Though I think a Sparse RSVD could be useful for cuML overall, especially as a primitive for implementing algorithms like TSVD and enabling LSA, I would follow sklearn's T-SNE implementation and just do a PCA on the input matrix X. This allows us to compare apples-to-apples with sklearn and will be less likely to confuse users when they are running the defaults.

Reply:
I agree with Corey.

Reply:
Ok cool :) I will definitely just use PCA(X) then as the initialization step :) [Probs through the Python layer].

But ye I'll also move the Sparse SVD to another PR then